test/performance/vectorization/unrolling/dotProduct.chpl

```diff
@@ -0,0 +1,209 @@
+//
+// This test has a series of functions that compute a dot product of two vectors.
+//
+
+use Time only stopwatch;
+use Random only randomStream;
+use CTypes only c_array;
+
+
+config const N = 10_000;
+config const iterations = 50;
+
+config type elementType = real(64);
+config param unrollFactor = 4;
+
+config const seed = 314159;
+config const minRand: elementType = -100;
+config const maxRand: elementType = 100;
+config const correctness = false;
+
+assert(N % unrollFactor == 0, "N must be a multiple of unrollFactor");
+compilerAssert(unrollFactor == 2 || unrollFactor == 4 || unrollFactor == 8,
+               "unrollFactor must be 2, 4, or 8");
+
+// Testing a simple serial for dot product
+proc dotProdFor(n: int, x: [1..n] elementType, y: x.type): elementType {
+  var s: elementType;
+  for i in 1..n do s += x[i] * y[i];
+  return s;
+}
+// Same as `dotProdFor`, but using a foreach loop
+proc dotProdForeach(n: int, x: [1..n] elementType, y: x.type): elementType {
+  var s: elementType;
+  foreach i in 1..n with (ref s) do s += x[i] * y[i];
+  return s;
+}
+// The Chapeltastic approach, using high level Chapel features
+proc dotProdChapeltastic(n: int, x: [1..n] elementType, y: x.type): elementType {
+  var s: elementType = + reduce (x * y);
+  return s;
+}
+// An unrolled version of the dot product, using slices
+proc dotProdSlices(n: int, x: [1..n] elementType, y: x.type): elementType {
+  var sv : [0..<unrollFactor] elementType;
+  foreach i in 1..(n & ~(unrollFactor - 1)) by unrollFactor do
+    sv += x[i..#unrollFactor] * y[i..#unrollFactor];
```

Comment:
Note that this is not unrolling and artificially heavy weight. This line here is essentially

```chpl
foreach ...
  forall idx in i..#unrollFactor with (+ reduce sv) {
    sv += x[idx] * y[idx]
  }
```

Same applies to all slice operations in this function.

---

I guess one could also make the argument that some promotions with slices in them should result in unrolling and not loops. It is probably too future-looking of an optimization, though.